runtime/vm/CRIUHelpers.cpp

```diff
@@ -536,4 +537,110 @@ addInternalJVMClassIterationRestoreHook(J9VMThread *currentThread, classIteratio
 	VM_CRIUHelpers::addInternalJVMClassIterationRestoreHook(currentThread, hookFunc);
 }
 
+
+jobject
+getRestoreSystemProperites(J9VMThread *currentThread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	jobject returnProperties = NULL;
+
+	if (NULL != vm->checkpointState.restoreArgsList) {
+		JavaVMInitArgs *restorArgs = vm->checkpointState.restoreArgsList->actualVMArgs;
+		 J9CmdLineOption* j9Options = vm->checkpointState.restoreArgsList->j9Options;
+		J9MemoryManagerFunctions * mmfns = vm->memoryManagerFunctions;
+		J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+		PORT_ACCESS_FROM_JAVAVM(vm);
+
+		UDATA count = 0;
+
+		for (IDATA i = 0; i < restorArgs->nOptions; ++i) {
+			char * optionString = restorArgs->options[i].optionString;
+
+			if (strncmp("-D", optionString, 2) == 0) {
+				count++;
+			}
+		}
+
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+
+		UDATA arrayEntries = count * 2;
+		j9object_t newArray = mmfns->J9AllocateIndexableObject(currentThread, ((J9Class*) J9VMJAVALANGSTRING_OR_NULL(vm))->arrayClass, arrayEntries, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+
+		if (NULL == newArray) {
+			vmFuncs->setHeapOutOfMemoryError(currentThread);
+			goto done;
+		}
+
+		for (IDATA i = 0, index = 0; i < restorArgs->nOptions; ++i, ++index) {
+			char * optionString = restorArgs->options[i].optionString;
+
+			if (strncmp("-D", optionString, 2) == 0) {
+				char* propValue = NULL;
+				char* propValueCopy = NULL;
+				char* propNameCopy = NULL;
+				UDATA propNameLen = 0;
+
+				propValue = strchr(optionString + 2, '=');
+				if (propValue == NULL) {
+					propNameLen = strlen(optionString) - 2;
+					propValue = optionString + 2 + propNameLen;
+				} else {
+					propNameLen = propValue - (optionString + 2);
+					++propValue;
+				}
+
+				UDATA valueLength = strlen(propValue);
+
+				propNameCopy = (char *)getMUtf8String(vm, optionString + 2, propNameLen);
+				if (NULL == propNameCopy) {
+					vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+					goto done;
+				}
+				propValueCopy = (char *)getMUtf8String(vm, propValue, valueLength);
+				if (NULL == propValueCopy) {
+					j9mem_free_memory(propNameCopy);
+					vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+					goto done;
+				}
+
+				j9object_t newObject = mmfns->j9gc_createJavaLangString(currentThread, (U_8 *)propNameCopy, propNameLen, J9_STR_TENURE);
```

Comment:
The `newArray` needs to be stored in a jni local or pushed /popped in a special frame around later allocates

Reply:
thanks, forgot about this